Convert "Deterministic builds with clang and lld" to Markdown

DavidSpickett opened the pull request:
Fixes #52

Sort of. Articles pre-sometime in 2020 are in HTML form and with https://github.com/JohannesKaufmann/html-to-markdown they convert pretty well but not enough to mass convert them.

But since I put the effort into trying it I figured I would clean up this one that had been (rightly) complained about.

I used a script to reformat the header, html2markdown to convert and then added back some plain text/monospace/code markers where they had been missed, and removed some escapes that got added because those plaintext bits weren't marked properly.

Now that it's Markdown, Hugo's theme will apply and produce the right colours for light and dark mode. The URL of the post will be the same.

DavidSpickett commented:
The orginal author is Nico Weber and I can't add reviewers here, or @ them, so I'll send them an email instead.

asl commented:
@nico

nico commented:
Thanks for the PR!

Is there a way to preview how the md file will appear on the blog?

asl commented:
@nico Yes, you'd need to clone the repo, and then run the local hugo instance.

DavidSpickett commented:
https://github.com/llvm/llvm-blog-www?tab=readme-ov-file#building-locally

nico commented:
> https://github.com/llvm/llvm-blog-www?tab=readme-ov-file#building-locally

I tried following that and got `ERROR deprecated: .Site.Author was deprecated in Hugo v0.124.0 and will be removed in Hugo 0.140.0. Implement taxonomy 'author' or use .Site.Params.Author instead.`. Maybe https://github.com/llvm/llvm-blog-www?tab=readme-ov-file#building-locally could mention which version of hugo should be used?

nico commented on content/posts/2019-11-07-deterministic-builds-with-clang-and-lld.md:
```diff
@@ -0,0 +1,166 @@
+
+For cache invalidation, consider using something finer-grained, such as only the latest commit of the directory containing the cache handling code, or the hash of all source files containing the cache handling code.
+
+For `--version output`, if your build is fully deterministic, the hash of the binary itself (and its dynamic library dependencies) can serve as a stable version identifier. You can keep a map of binary hash to all commit hashes that produce that binary somewhere.
```
This should be:

```
For `--version` output
```

(ie "output" should not be part of the code block)

DavidSpickett replied:
done.

nico commented on content/posts/2019-11-07-deterministic-builds-with-clang-and-lld.md:
```diff
@@ -0,0 +1,166 @@
+
+If you don’t want to require devs to cd into the build directory for debugging to work, you have to do debugger-specific configuration tweaks.
+
+To make sure devs don’t miss this, you could have your custom init script set an env var and query if it’s set early during your test binary startup, and exit with a message like “Add `source /path/to/your/project/gdbinit` to your ~/.gdbinit” if the environment variable isn’t set.
```
`~/.gdbinit` should be a code block

DavidSpickett replied:
Done.

nico commented on content/posts/2019-11-07-deterministic-builds-with-clang-and-lld.md:
```diff
@@ -0,0 +1,166 @@
+
+1. Run `subst X: C:\src\real\root` in cmd.exe before launching the debuggers to create a virtual drive that maps X: to the actual source location. Both windbg and Visual Studio will load code over X: this way.
+2. Add `C:\src\real\root` to each debugger’s source search path.
+3. - [Windbg](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/source-window#source-path): Run `.srcpath+ C:\src\real\root`. You can also set this via the `_NT_SOURCE_PATH`  environment variable, or via  File->Source File Path (Ctrl+P). Or pass `-srcpath C:\src\real\root` when launching windbg from the command line.
```
This should be a `3.`, it's a sub-list of `2.`

DavidSpickett replied:
Will fix, this is one thing the converter gets wrong, then I got it wrong manually fixing it.

DavidSpickett replied:
Done.

nico reviewed (commented):
Thank you very much for doing this! This looks good except for a few small things below, and for one larger issue: IIRC I added the yellow highlighting back then to highlight all the flags and important parts that one should use, so that it's easy to scroll quickly through the document and see the "what should I actually do" bits. I don't know if we have something for that.

I suppose markdown has `**foo**` for highlighting, and maybe hugo makes it possible to set background-color on whatever class strong in markdown generates, and that could be set depending on light or dark mode? I don't know how that's usually done. (e.g. "Do not use these flags" was marked for emphasis, as were a bunch of flags.)

More of a question than a comment: It looks like this does one line of text per paragraph. Is that how all the md files look in this repo?

Also, I've been meaning to update the post with some new things for a long time, but it's good to do a clean conversion without any other changes first.

DavidSpickett commented:
> I got it working locally with this diff

I'm using Hugo v0.119.0 so I didn't see this issue. There hasn't been a change to the theme in a long time (https://github.com/avianto/hugo-kiera) so I think we'd be ok patching it in this repo instead.

I can do that if you like, if you don't mind me starting from your diff?

> IIRC I added the yellow highlighting back then to highlight all the flags and important parts that one should use, so that it's easy to scroll quickly through the document and see the "what should I actually do" bits. I don't know if we have something for that.

We can use HTML tags in Markdown, as long as we choose one that will be readable in a dark theme. I see there is a `<mark>Marked text</mark>`. I will try that.

> More of a question than a comment: It looks like this does one line of text per paragraph. Is that how all the md files look in this repo?

I think this is just what the converter decided to do. Markdown can be same line or multiple and it'll treat it as one either way.

I can 80 column break it if you like, but perhaps I do that once I've fixed the other things?

nico commented:
> I can do that if you like, if you don't mind me starting from your diff?

Sure, thanks :) That was just local hacks to get things building though. I'm guessing we might not want to remove GoogleAnalytics on the actual page? (Don't know though.)

> I think this is just what the converter decided to do. Markdown can be same line or multiple and it'll treat it as one either way.
> 
> I can 80 column break it if you like, but perhaps I do that once I've fixed the other things?

No strong opinion, do whatever you think works best in this repo.

Maybe git diffs look a bit better (including on github) if paragraphs are split over several lines, but up to you.

DavidSpickett commented:
> We can use HTML tags in Markdown, as long as we choose one that will be readable in a dark theme. I see there is a <mark>Marked text</mark>. I will try that.

I've done this, it works well for light and dark themes.

Except for the ml.py link which comes out white link text on yellow background, so I've highlighted the text that follows it instead.

There is also a note at the start so the reader knows why bits of the text are formatted this way (it was not obvious to me at least).

DavidSpickett commented:
> No strong opinion, do whatever you think works best in this repo.

I had a go at it but I think it comes down to what's best for the person editing it. An 80 column wrap is a waste of space on a wide screen and the rendered copy will be automatically wrapped anyway. You're the author so you can wrap it later if it helps you make changes.

DavidSpickett commented:
2 ways to be compatible with latest Hugo versions:
https://github.com/llvm/llvm-blog-www/pull/64
https://github.com/llvm/llvm-blog-www/pull/65

nico reviewed (approved):
Thanks much!

DavidSpickett commented:
@asl can this be merged now?
